tools/mc-java-protoc/src/main/java/io/spine/tools/mc/java/protoc/EntityMatcher.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2021, TeamDev. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.tools.mc.java.protoc;
+
+import io.spine.tools.protoc.FilePattern;
+import io.spine.tools.protoc.ForEntities;
+import io.spine.tools.protoc.ProtoOption;
+import io.spine.type.MessageType;
+
+import java.util.List;
+import java.util.function.Predicate;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A predicate which checks if a {@link MessageType} represents an entity state type.
+ *
+ * <p>An entity state type may be marked by a Protobuf option or be matched by a file pattern.
+ */
+public final class EntityMatcher implements Predicate<MessageType> {
+
+    private final ForEntities entities;
+
+    public EntityMatcher(ForEntities entities) {
+        this.entities = checkNotNull(entities);
+    }
+
+    @Override
+    public boolean test(MessageType type) {
+        List<ProtoOption> options = entities.getOptionList();
+        boolean match = options.stream()
+                               .map(OptionMatcher::new)
+                               .anyMatch(matcher -> matcher.test(type));
+        if (match) {
+            return true;
+        }
+        List<FilePattern> patterns = entities.getPatternList();
```

Comment:
Please extract two methods, one for checking agains options, and another for checking against patterns.

Reply:
I've refactored this class into a more functional style. Please have a look. There's definitely less duplicate now. I want to see how readable is it to other people.